### DISCUSSION TITLE
Guide needed on using queues

### DISCUSSION THREAD

**Discussion by waptik** [oliver-oloughlin/kvdex#113] [General]
So i haven't used the native queuing system(`Deno.enqueue`, `Deno.listenQueue`) before. But i'd like to see a realistic example of its usage both at the model(`db.users.enqueue`, `db.users.listenQueue`, `db.users.deleteUndelivered`, `db.users.findUndelivered`) and database(`db.enqueue`, `db.listenQueue`) levels in a live server such as `Deno.serve` or `oakjs`.
In the readme file, the model structure for users wasn't defined when it came to using queue related ops

If an example is not possible, a guide as to where to use what would be very much appreciated.

**Comment by waptik** [comment]
I tried out `enqueue/listenQueue` at the model level and found out that the data has nothing to do with the model's schema(no type error or runtime error).

**Reply by oliver-oloughlin** [reply]
This is the correct behaviour. Enqueueing data, both at the database level and at the collection level, accepts any data type. The only difference is really which listeners the data goes to. Therefore, you can view the database/collection as a top-level topic. I chose this strategy because I did not want to limit any use cases. I myself saw little use of only allowing to send user objects in a user collection queue for example. The reason being I would often want to create some custom message object which would include the action that should be taken by the listener. I can give an example:

```ts
const db kvdex(kv, {
  users: collection(model<User>())
})

type UserQueueMessage = {
  action: "delete",
  id: KvId,
} | {
  action "add",
  user: User
}

db.users.listenQueue<UserQueueMessage>(async (msg) => {
  if (msg.action === "delete") {
    await db.users.delete(msg.id)
  } else if (msg.action === "add") {
    await db.users.add(msg.users)
  }
})
```

**Reply by waptik** [reply]
I spent some time playing around with queues at both database&collection levels and i can confirm that it makes it easy to know the data structure each listener is receiving. The addition of passing a `topic` to the queue gives it another superpower.

Any recommendation on where to place the listening queues to make things clean a bit or it doesn't really matter where?

Oh, i tried passing a type to a listening queue as you did in the example above but some type mismatch error occurred but no biggie!

Thanks.

![image](https://github.com/oliver-oloughlin/kvdex/assets/1687551/d97c6188-9a4a-4381-9248-75f1d80823d1)

**Reply by oliver-oloughlin** [reply]
I recommend placing queue listeners somewhere near the top level of your program structure, also near the top of the file they are located in. And also to make their placement logical by naming. That way it makes it easier to have an overview of the active listeners you have. It can quickly become messy if you have many listeners spread across all sorts of different locations.

The type error I believe is because you are using an interface, and it's not explicity extending the QueueValue type. Although it can't actually extend the QueueValue type directly, as it's not strictly an object type, you can extend the KvObject type which is one of the valid QueueValue types:
```ts
import { kvdex, KvObject } from "./mod.ts"

const kv = await Deno.openKv()
const db = kvdex(kv, {})

interface IQueueData extends KvObject {
  msg: {
    id: number
    chat: number
    from: number
  }
  data: string
}

// No type error
db.listenQueue<IQueueData>((msg) => console.log(msg))
```

Note that you could also simply use a type instead of an interface and it would work just the same.

**Reply by waptik** [reply]
Alright thanks. I put them together in a file mainly made for queues and imported that file like this `import './queues.ts'` in my main file.

Also i think i'm starting to like using `type` instead of `interface` when defining type definitions for my queues data. It saves me the need to import `KvObject`.

Thanks again!

**Reply by waptik** [reply]
It seems something has changed again with using `type` in v0.29.1(had to upgrade because `kvdex` was returning `any` in v0.28.3 so i wasn't getting type definitions for `db.*`).
![image](https://github.com/oliver-oloughlin/kvdex/assets/1687551/97344284-1574-4d6e-a319-90bff496cbf3)
Also, using `interface` by extending the `KvObject` type, vscode is screaming at me
![image](https://github.com/oliver-oloughlin/kvdex/assets/1687551/90cd0456-0888-4999-84ad-e2b234d9569a)
![image](https://github.com/oliver-oloughlin/kvdex/assets/1687551/dd427c3e-73a0-44a7-87d1-715ef5680209)

**Reply by oliver-oloughlin** [reply]
Have you remembered to cache your dependencies after updating the version?

The interface issue I would guess is because any included interfaces must also extend KvObject, meaning User must extend KvObject.

**Reply by waptik** [reply]
Jeez, i forgot to do that but it didn’t help either. I initially started
with  “deno cache —reload dev.ts” then moved on to using the command
palette to do the same thing but it didn’t seem to fix the problem.
Oh okay, too much trouble when using interfaces 🤦🏾‍♂️. I’ll probably have
to stick using interface for now even though there’ll be some extra work.

**Update**: after extending User interface with KvObject, it seems to have also fixed the issue with "type".
![image](https://github.com/oliver-oloughlin/kvdex/assets/1687551/394a9058-e5d4-4453-a1e6-648a72ff470b)

**Reply by oliver-oloughlin** [reply]
Re-caching like you described should have worked. Another alternative that I like to use is setting this in your /.vscode/settings.json:
```json
{
  "deno.cacheOnSave": true,
}
```

If you are still having issues with the DB types, would you mind showing how the database schema is defined?

I pretty much always use types because in my opinion they are better in almost every use case, and can be used for everything that interfaces are used for. The issue you were facing is my main reason for avoiding interfaces: TypeScript doesn't recognize an interface as the correct type before it actually extends the specific type, even if the interface itself conforms to the type. Meanwhile TypeScript uses duck-typing for types, meaning if the signature is correct then it sees it as a matching type.

**Reply by waptik** [reply]
So i used to have `cacheOnSave` set to `true` but i had to disable it because i've been facing [ this issue](https://github.com/denoland/vscode_deno/issues/943)
Oh okay, that's some nice understanding of things. There is little solid information about the differences between interfaces and types and most seem to say types and interfaces can be used interchangeably.

Here's a portion of the client definition and that of some collections

```typescript

//------------------------------ collection schema defintions inside db/models.ts ------------------------------//
import { z } from "zod";

// @see https://stackoverflow.com/a/73825370
const withDefautTimestamps = z.object({
  created_at: z
    .string()
    .datetime()
    .default(() => new Date().toISOString()),
  updated_at: z
    .string()
    .datetime()
    .default(() => new Date().toISOString()),
});

export type IDBAudioVoice = z.infer<typeof AudioVoiceSchema>;

....

export const AudioVoiceSchema = z.object({
  // information from telegram
 ...
  file_unique_id: z.string().describe("unique"),

  key: z.string().describe("unique"),
  ...

  uploaded_by: z.number(),
  message_id: z.number(),
}).merge(withDefautTimestamps);
....

//------------------------------ db client definition in db/client.ts ------------------------------//
const dbClient = kvdex(kv, {
  users: collection(zodModel(UserSchema), {
    indices: {
      telegram_user_id: "primary", // unique
      id: "secondary", // non-unique
    },
    // history: true,
  }),
// some omitted
  audio_voices: collection(zodModel(AudioVoiceSchema), {
    indices: {
      file_unique_id: "primary", // unique
      file_id: "secondary", // non-unique
      key: "secondary", // non-unique
    },
    // history: true,
  }),
});
```

**Comment by waptik** [comment]
Hi again,

So i've come to face another issue regarding processing queues with `.collection.listenQueue` inside a fresh app. I'm not sure if it's a bug from fresh or kvdex so i'm also referencing the opened issue from fresh's repo here as well.
Ref: https://github.com/denoland/fresh/issues/2186

**Reply by oliver-oloughlin** [reply]
I see. I am happy to help anyway I can, and have a few thoughts:

1. I would start by trying to use the native `kv.listenQueue()` to see if there's any difference, just to identify if it is a kvdex specific issue or not.
2. Make sure you are invoking `listenQueue()` and `enqueue()` on the same collection (collection queues are treated as their own topics).
3. When are you running `enqueue()` compared to `listenQueue()`? I experienced an issue a little while back where if I fired an `enqueue()` very fast after `listenQueue()`, the queue lsitener had not been registered/created yet, and therefore didn't catch the message.
4. Make sure the data being queued is not undefined, this will lead to it not being delivered as expecetd due to how kvdex parses queue messages to deliver them to the correct topics.

**Reply by waptik** [reply]
1. As mentioned in the referenced issue, it wasn't specific to kvdex because the `kv.listenQueue` had no effect.
2. In the screenshots provided on the other side, you could see that they were being invoked on the same collection
3. `enqueue()` is being fired on-demand so it's after `listenQueue()` as it's defined in entrypoint file of fresh (`main.ts`)
4. I can assure you that the data passed to `enqueue()` is always defined as i made sure not to queue them if they are undefined
I'm going to upload the same pics here again

- kvdex `.collection.enqueue()` and `.collection.listenQueue()`
![screen-2023-12-20-06-56-35](https://github.com/oliver-oloughlin/kvdex/assets/1687551/4ca6bd9f-33a3-43a7-8a99-f1614a6ff8ae)
![screen-2023-12-20-07-00-59](https://github.com/oliver-oloughlin/kvdex/assets/1687551/03f176dd-c70c-403d-81c1-11a4629df1ef)

- Native `kv.enqueue()` and `kv.listenQueue()`
![screen-2023-12-20-07-08-55](https://github.com/oliver-oloughlin/kvdex/assets/1687551/2c6e6b88-48fd-4c4a-84fd-a6b2937198e5)
![screen-2023-12-20-07-09-46](https://github.com/oliver-oloughlin/kvdex/assets/1687551/18e94209-2cc9-4957-ad4d-af082bc974f2)

**Comment by waptik** [comment]
I just deployed the [example repo](https://github.com/waptik/fresh-kvdex-queues) to deploy and it's working as it should. AS you can see from the screenshot below, `listenQueue` from both native kv and kvdex are able to receive data effectively so i think this is a fresh issue in local development with the `dev.ts` file
![image](https://github.com/oliver-oloughlin/kvdex/assets/1687551/257fde1e-899c-4118-931b-7fad46bea773)

Thanks for your help!

**Reply by oliver-oloughlin** [reply]
Yeah, seems like it is actually an issue with fresh itself. Good job investigating it. I'll follow any developments in the issue you posted. I can't really help much with issues that come from fresh, but I'll keep an eye out for anything that needs to be changed with kvdex.

**Reply by waptik** [reply]
Sure, thanks again.

**Comment by ryanleecode** [comment]
I still don't understand why would we need queues for a specific model if the callback from `listenQueue` does not include a value from the model itself?

**Reply by oliver-oloughlin** [reply]
Really, it's just a matter of logical structuring. Collection-specific queues can be seen as their own topic.
If you were for example to have a lot of listeners doing tasks specific to a collection, it might be simpler to use collection queues instead of many top-level db queues with custom-set topics which require you to give the correct topic when enqueueing messages as well.
